### PR TITLE
SEQNG-1046 Implemented GMOS N&S FITS keywords.

### PR DESCRIPTION
**Pull request by jluhrs**

### CODE REVIEW

File: modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosKeywordReader.scala [hunk 1]
```diff
@@ -26,18 +32,66 @@ final case class GmosObsKeywordsReader[F[_]: Sync](config: Config) {
...
       .map(_.toBoolean)
       .safeValOrDefault
+
+  def nodMode: F[String] = "STANDARD".pure[F]
+
+  def nodPix: F[Int] = Gmos.nodAndShuffle(config).map(_.rows).explainExtractError
+
+  def nodCount: F[Int] = Gmos.nodAndShuffle(config).map(_.cycles).explainExtractError
+
+  def nodAxOff: F[Double] = Gmos.nodAndShuffle(config).explainExtractError
```
Comment (tpolecat) [inline modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosKeywordReader.scala]: These four methods are identical other than the stage and angle selector. I think you forgot to change the error message in the last two when you copy/pasted. Can we factor out one method and call it with different arguments?

File: modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosHeader.scala [hunk 2]
```diff
@@ -96,21 +122,23 @@ object GmosHeader {
             buildString(gmosReader.dcName, KeywordName.GMOSDC),
             buildString(gmosReader.detectorType, KeywordName.DETTYPE),
             buildString(gmosReader.detectorId, KeywordName.DETID),
-            buildDouble(gmosReader.exposureTime, KeywordName.EXPOSURE),
+            buildDouble(
+              gmosObsReader.isNS.ifM(
+                (gmosObsReader.nodCount, gmosReader.exposureTime).mapN{ case (c, t) => 2 * c * t },
```
Comment (cquiroz) [inline modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosHeader.scala]: I tend to put these calculations on the keyword reader and leave the `Header` class just as set of values and keyword names. WDYT?

File: modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosKeywordReader.scala [hunk 3]
```diff
@@ -26,18 +32,66 @@ final case class GmosObsKeywordsReader[F[_]: Sync](config: Config) {
...
+        .getOrElse(SeqexecFailure.Unexpected("Cannot find stage A parameters in step configuration.")
+          .raiseError[F, Double]
+        )
+    )
+
+  def nodBxOff: F[Double] = Gmos.nodAndShuffle(config).explainExtractError
+    .flatMap(
+      _.positions.find(_.stage === StageB)
+        .map(x => (x.offset.p.toAngle.toMicroarcseconds.toDouble/1e6).pure[F])
+        .getOrElse(SeqexecFailure.Unexpected("Cannot find stage A parameters in step configuration.")
```
Comment (cquiroz) [inline modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosKeywordReader.scala]: stage B?

File: modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosKeywordReader.scala [hunk 4]
```diff
@@ -26,18 +32,66 @@ final case class GmosObsKeywordsReader[F[_]: Sync](config: Config) {
...
+        )
+    )
+
+  def isNS: F[Boolean] = config.extractInstAs[java.lang.Boolean](USE_NS_PROP).map(_.booleanValue).explainExtractError
+
+}
+
+object GmosObsKeywordsReader {
+
+  private implicit class ExplainExtractError[F[_]: Sync, A](v: Either[ExtractFailure, A]) {
```
Comment (cquiroz) [inline modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosKeywordReader.scala]: Do you need `Sync` here? I think `MonadError` is enough

File: modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosHeader.scala (hunk 2 above)
Comment (jluhrs) [reply modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosHeader.scala]: I agree in general, but the problem here is that the calculation combines values from two readers.

File: modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosKeywordReader.scala (hunk 4 above)
Comment (jluhrs) [reply modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosKeywordReader.scala]: Y tried with `MonadError` first, but then I got compilation errors about ambiguous implicits.

File: modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosHeader.scala (hunk 2 above)
Comment (cquiroz) [reply modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosHeader.scala]: Good point

File: modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosKeywordReader.scala (hunk 4 above)
Comment (cquiroz) [reply modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosKeywordReader.scala]: That may mean you are missing to explicity set the `F`. e.g. `widenRethrowT[F]`?

Comment (jluhrs) [reply modules/seqexec/server/src/main/scala/seqexec/server/gmos/GmosKeywordReader.scala]: I rearranged the type parameters and put some explicitly, and now it compiles with `MonadError`

**Pull request merged**